src/core/net/ip6.cpp

```diff
@@ -962,17 +962,20 @@ const NetifUnicastAddress *Ip6::SelectSourceAddress(MessageInfo &aMessageInfo)
     {
         candidateId = netif->GetInterfaceId();
 
+        if (destination->IsLinkLocal() || destination->IsMulticast())
+        {
+            if (interfaceId != candidateId)
+            {
+                continue;
+            }
+        }
+
         for (const NetifUnicastAddress *addr = netif->mUnicastAddresses; addr; addr = addr->GetNext())
         {
             candidateAddr = &addr->GetAddress();
+            uint8_t destinationScope = (destination->PrefixMatch(*candidateAddr) >= addr->mPrefixLength) ?
```

Comment:
Also, `destination->PrefixMatch(*candidateAddr)` is called again below.  Would be good to store the result in a local variable.